### PR TITLE
feat(ast): impl PartialEq, Eq for Expr

### PR DESCRIPTION
**Pull request by yuanbohan**
## what's included

- use enum to refactor ast instead of optional fields to match
- implement PartialEq, Eq trait for the ease of comparison
- some `Expr::new_` functions, which will be used in `promql.y` in next PR

### CODE REVIEW

**Comment by yuanbohan** [comment]
@evenyag @waynexia PTAL

File: src/parser/ast.rs [hunk 1]
```diff
@@ -140,64 +258,246 @@ pub enum Expr {
...
-}
-
-impl Display for VectorMatchCardinality {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        match *self {
-            VectorMatchCardinality::OneToOne => write!(f, "one-to-one"),
-            VectorMatchCardinality::ManyToOne => write!(f, "many-to-one"),
-            VectorMatchCardinality::OneToMany => write!(f, "one-to-many"),
-            VectorMatchCardinality::ManyToMany => write!(f, "many-to-many"),
+    pub fn new_aggregate_expr(
```
Comment (yuanbohan) [inline src/parser/ast.rs:415]: https://github.com/prometheus/prometheus/blob/main/promql/parser/parse.go#L378

File: src/parser/ast.rs [hunk 2]
```diff
@@ -140,64 +258,246 @@ pub enum Expr {
...
+                Some(_) => already_set_err,
+            },
+            _ => {
+                Err("@ modifier must be preceded by an instant vector selector or range vector selector or a subquery".into())
+            }
+        }
+    }
+
+    /// set offset field for specified Expr, but CAN ONLY be set once.
+    pub fn offset_expr(self, offset: Offset) -> Result<Self, String> {
```
Comment (yuanbohan) [inline src/parser/ast.rs:366]: https://github.com/prometheus/prometheus/blob/main/promql/parser/parse.go#L686

File: src/parser/ast.rs [hunk 3]
```diff
@@ -140,64 +258,246 @@ pub enum Expr {
...
+                    range,
+                });
+                Ok(ms)
+            }
+            _ => Err("ranges only allowed for vector selectors".into()),
+        }
+    }
+
+    /// set at_modifier for specified Expr, but CAN ONLY be set once.
+    pub fn step_invariant_expr(self, at_modifier: AtModifier) -> Result<Self, String> {
```
Comment (yuanbohan) [inline src/parser/ast.rs:335]: https://github.com/prometheus/prometheus/blob/main/promql/parser/parse.go#L721-L791

File: src/parser/ast.rs [hunk 4]
```diff
@@ -140,64 +258,246 @@ pub enum Expr {
...
+    pub fn new_number_literal(val: f64) -> Result<Self, String> {
+        Ok(Expr::NumberLiteral(NumberLiteral { val }))
+    }
+
+    pub fn new_string_literal(val: String) -> Result<Self, String> {
+        Ok(Expr::StringLiteral(StringLiteral { val }))
+    }
+
+    /// NOTE: @ and offset is not set here.
+    pub fn new_matrix_selector(expr: Expr, range: Duration) -> Result<Self, String> {
```
Comment (yuanbohan) [inline src/parser/ast.rs:315]: https://github.com/prometheus/prometheus/blob/main/promql/parser/generated_parser.y#L423-L446

File: src/parser/ast.rs [hunk 5]
```diff
@@ -35,47 +137,52 @@ pub struct EvalStmt {
...
 pub struct AggregateExpr {
-    pub op: TokenType,         // The used aggregation operation.
-    pub expr: Box<Expr>,       // The Vector expression over which is aggregated.
-    pub param: Box<Expr>,      // Parameter used by some aggregators.
-    pub grouping: Vec<String>, // The labels by which to group the Vector.
-    pub without: bool,         // Whether to drop the given labels rather than keep them.
+    pub op: TokenType,            // The used aggregation operation.
+    pub expr: Box<Expr>,          // The Vector expression over which is aggregated.
+    pub param: Option<Box<Expr>>, // Parameter used by some aggregators.
+    pub grouping: AggModifier,
```
Comment (waynexia) [inline src/parser/ast.rs:150]: Putting those tailing comments above its fields can make them rendered as field documentation. E.g.:
``` rust
/// The used aggregation operation.
pub op: TokenType
```

File: src/parser/ast.rs [hunk 6]
```diff
@@ -13,11 +13,113 @@
...
 #![allow(dead_code)]
-use std::fmt::{self, Display};
+use crate::label::Matchers;
+use crate::parser::token::{self, T_END, T_START};
+use crate::parser::{Function, FunctionArgs, Token, TokenType};
 use std::time::{Duration, SystemTime};
 
-use crate::label::Matchers;
-use crate::parser::{Function, TokenType};
+// TODO: better PartialEq, Eq ignoring Vec<String> orders
```
Comment (waynexia) [inline src/parser/ast.rs:21]: How about also changing `Vec` to `HashSet` like `Matchers`

**Review by waynexia — APPROVED** [review]
Brilliant! I haven't looked into the details but it generally looks good to me 👍

File: src/parser/ast.rs (hunk 5 above)
Comment (yuanbohan) [reply src/parser/ast.rs:150]: :ok_hand:

File: src/parser/ast.rs (hunk 6 above)
Comment (yuanbohan) [reply src/parser/ast.rs:21]: :ok_hand:

**Pull request merged**